ddtelemetry/src/worker/mod.rs

```diff
@@ -454,10 +454,9 @@ impl TelemetryWorker {
                 });
                 future::join_all(futures).await;
 
-                let mut self_lock = self_arc.write().await;
-                self_lock.data.started = false;
-                if !self_lock.config.restartable {
```

Comment:
I can be mistaken, but I thought that the use of the the `RwLock` here was about ordering and visibility guarantees. But maybe that's being taken care of by the `await`.

Reply:
The `RwLock` is about the fact that both futures may try to write into self, and they may run in parallel (if there was a simple loop instead of `join_all` this wouldn't be the case, because futures run only when polled). If not, this would have sufficed (with the creation of a shared reference `self2`, which is `Copy`, unlike the mut ref `self`)

```
let self2: &TelemetryWorker = &*self;
let futures = payloads.into_iter().map(|payload| {
	async move {
		match self2.send_payload(&payload).await {
			Ok(()) => (),
			Err(err) => self2.log_err(&err),
		}
	}
});
```

I'm pretty sure there are no visibility problems; this would even work with plain threads because of the synchronization guarantees of a join.

Reply:
Yes, I know what the `RwLock` is used for inside the futures (and how `rust` futures work). I just wasn't sure what guarantees the `join` will give you, but since any writes happens before joining (which needs to make things visible) it sounds fine.